src/evergreen/async_api.py

```diff
@@ -0,0 +1,1443 @@
+from evergreen.task_annotations import TaskAnnotation
+from evergreen.task_reliability import TaskReliability
+from evergreen.tst import Tst
+from evergreen.users_for_role import UsersForRole
+from evergreen.util import evergreen_input_to_output, format_evergreen_date
+from evergreen.version import RecentVersions, Requester, Version
+
+structlog.configure(logger_factory=LoggerFactory())
+LOGGER = structlog.getLogger(__name__)
+
+CACHE_SIZE = 5000
+DEFAULT_LIMIT = 100
+MAX_RETRIES = 3
+START_WAIT_TIME_SEC = 2
+MAX_WAIT_TIME_SEC = 5
+
+
+@dataclass(eq=True, frozen=True)
+class EvergreenResponse:
+    """Container object for parsing responses."""
+
+    next_url: Optional[str] = None
+    text: Optional[str] = None
+    json: Optional[Union[List[Dict[str, Any]], Dict[str, Any]]] = None
+
+
+class AsyncEvergreenApi(EvergreenApi):
+    """Base methods for building API objects."""
+
+    def __init__(
+        self,
+        api_server: str = DEFAULT_API_SERVER,
+        auth: Optional[EvgAuth] = None,
+        timeout: Optional[int] = None,
+        session: Optional[ClientSession] = None,
+        log_on_error: bool = False,
+    ) -> None:
+        """
+        Create a _BaseEvergreenApi object.
+
+        :param api_server: URI of Evergreen API server.
+        :param auth: EvgAuth object with auth information.
+        :param timeout: Time (in sec) to wait before considering a call as failed.
+        :param session: Session to use for requests.
+        :param log_on_error: Flag to use for error logs.
+        """
+        super().__init__(api_server, auth, timeout, None, log_on_error)
+        self._async_session = session
+
+    @asynccontextmanager
+    async def with_async_session(self) -> AsyncGenerator["AsyncEvergreenApi", None]:
+        """Yield an instance of the API client with a shared session."""
+        headers = {}
+        auth = self._auth
+        if auth:
+            headers["Api-User"] = auth.username
+            headers["Api-Key"] = auth.api_key
+        async with self.async_session() as session:  # type: ignore[attr-defined,misc]
+            api = AsyncEvergreenApi(self._api_server, self._auth, self._timeout, session)
+            yield api
+
+    @asynccontextmanager
+    async def async_session(self) -> AsyncGenerator[ClientSession, None]:
+        """
+        Get the shared session if it exists, else create a new session.
+
+        :return: Session to query the API with.
+        """
+        if self._async_session:
+            yield self._async_session
+        else:
+            yield await self._create_async_session()
+
+    async def _create_async_session(self) -> ClientSession:
+        """Create a new session to query the API with."""
+        headers = {}
+        auth = self._auth
+        if auth:
+            headers["Api-User"] = auth.username
+            headers["Api-Key"] = auth.api_key
+        return ClientSession(headers=headers)
+
+    def _create_url(self, endpoint: str) -> str:
```

Comment:
Any reason you needed to re-implement methods like this?

Reply:
cleaning this up